File: timeline/timeline.fs
module Timeline

type Timeline<'a> =
    { mutable _last: 'a
      mutable _fns: list<'a -> unit> }

let Timeline =
    fun a ->
        { _last = a
          _fns = [] }

module TL =
    let last =
        fun timeline ->
            timeline._last

    let next =
        fun a timeline ->
            timeline._last <- a // mutable
            timeline._fns
            |> List.iter (fun f -> f a) //perform all fns in the list

    let bind =
        fun monadf timelineA ->
            let timelineB =
                timelineA._last |> monadf
            let newFn =
                fun a ->
                    let timeline =
                        a |> monadf
                    timelineB
                    |> next timeline._last

            timelineA._fns <- timelineA._fns @ [ newFn ]
            timelineB

    let map =
        fun f timelineA ->
            let timelineB =
                Timeline (timelineA._last |> f)
            let newFn =
                fun a ->
                    timelineB
                    |> next (a |> f)

            timelineA._fns <- timelineA._fns @ [ newFn ]
            timelineB

    let unlink =
        fun timeline ->
            timeline._fns <- []

File: code/py/timeline.py
from __future__ import annotations
from typing import Callable, Generic, List, TypeVar, Optional

A = TypeVar('A')
B = TypeVar('B')

class Timeline(Generic[A]):
    def __init__(self, a: A):
        self._last: A = a
        self._fns: List[Callable[[A], None]] = []

    def last(self) -> A:
        return last(self)

    def next(self, a: A) -> None:
        next(self, a)

    def bind(self, monadf: Callable[[A], Timeline[B]]) -> Timeline[B]:
        return bind(self, monadf)

    def map(self, f: Callable[[A], B]) -> Timeline[B]:
        return map(self, f)

    def unlink(self) -> None:
        unlink(self)

def last(timeline: Timeline[A]) -> A:
    return timeline._last

def next(timeline: Timeline[A], a: A) -> None:
    timeline._last = a
    for f in timeline._fns:
        f(a)

def bind(timelineA: Timeline[A], monadf: Callable[[A], Timeline[B]]) -> Timeline[B]:
    timelineB: Timeline[B] = monadf(timelineA._last)

    def newFn(a: A) -> None:
        timeline: Timeline[B] = monadf(a)
        next(timelineB, timeline._last)

    timelineA._fns.append(newFn)
    return timelineB

def map(timelineA: Timeline[A], f: Callable[[A], B]) -> Timeline[B]:
    timelineB: Timeline[B] = Timeline(f(timelineA._last))

    def newFn(a: A) -> None:
        next(timelineB, f(a))

    timelineA._fns.append(newFn)
    return timelineB

def unlink(timeline: Timeline[A]) -> None:
    timeline._fns = []